File: Drawer.py
import argparse
import configparser
from datetime import datetime
import json

import cv2
import numpy as np
import requests
import time
import os

interpolation_modes = {
    "near": cv2.INTER_NEAREST,
    "lin": cv2.INTER_LINEAR,
    "cube": cv2.INTER_CUBIC,
    "area": cv2.INTER_AREA,
    "lanc4": cv2.INTER_LANCZOS4,
    "lin_ex": cv2.INTER_LINEAR_EXACT,
}


def getPixelWallData():
    global wall
    if cfg['network'] == "mainnet":
        url = "https://beaconcha.in/api/v1/graffitiwall"
    else:
        url = "https://" + cfg['network'] + ".beaconcha.in/api/v1/graffitiwall"
        return
    try:
        page = requests.get(url)
    except requests.exceptions.RequestException as _:
        print("can't reach graffitiwall")
        return
    if page.status_code != 200:
        print("error fetching wall")
        return
    w = page.json()["data"]

    # filter visible area
    wall = np.full((img.shape[0], img.shape[1], 3), 255, np.uint8)
    for pixel in w:
        if y_offset <= pixel["y"] < y_offset + y_res and \
                x_offset <= pixel["x"] < x_offset + x_res:
            wall[pixel["y"] - y_offset, pixel["x"] - x_offset] = tuple(
                int(pixel["color"][i:i + 2], 16) for i in (0, 2, 4))


def updateDrawPixels():
    # add already set pixels which might need to be re-drawn
    # TODO do this without loop? but the wall is full of wholes
    same = np.all(img[..., :3] == wall, axis=-1)
    return ~(same + transparent_pixels)


def getPixel():
    draw_y, draw_x = np.where(draw_pixels)
    if len(draw_y) > 0:
        random_pixel_index = np.random.choice(len(draw_y))
        y = draw_y[random_pixel_index]
        x = draw_x[random_pixel_index]
        color = format(img[y][x][0], '02x')
        color += format(img[y][x][1], '02x')
        color += format(img[y][x][2], '02x')
        return "gw:" + str(x + x_offset).zfill(3) + str(y + y_offset).zfill(3) + color
    return "graffiti done"


def getImage():
    global x_res, y_res
    file = cfg['ImagePath']
    if not os.path.isabs(file):
        file = os.path.dirname(os.path.abspath(__file__)) + "/" + file
    orig_img = cv2.imread(file, cv2.IMREAD_UNCHANGED)
    y_res, x_res, channels = orig_img.shape
    scale = int(cfg['scale'])
    if cfg['XRes'] != "original":
        x_res = int(cfg['XRes'])
    else:
        x_res = int(x_res * (scale / 100))
    if cfg['YRes'] != "original":
        y_res = int(cfg['YRes'])
    else:
        y_res = int(y_res * (scale / 100))
    int_mode = cfg["interpolation"]
    if int_mode not in interpolation_modes:
        print("unknown interpolation mode: " + cfg["interpolation"])
        exit(1)
    resized = cv2.resize(orig_img, dsize=(x_res, y_res), interpolation=interpolation_modes[int_mode])
    # support for various formats (grayscale, jpg, png, ...)
    if channels == 1:
        img = cv2.cvtColor(resized, cv2.COLOR_GRAY2RGBA)
    elif channels == 3:
        img = cv2.cvtColor(resized, cv2.COLOR_BGR2RGBA)
    elif channels == 4:
        img = cv2.cvtColor(resized, cv2.COLOR_BGRA2RGBA)
    return img, min(1000 - x_res, int(cfg['XOffset'])), min(1000 - y_res, int(cfg['YOffset']))


def setNimbusGraffiti(graffiti):
    url = "http://" + args.eth2_url + ":" + str(args.eth2_port) + "/api/nimbus/v1/graffiti"
    header = {'content-type': 'text/plain'}

    try:
        response = requests.post(url, headers=header, data=graffiti)
    except requests.exceptions.RequestException as e:
        return False
    return response.ok


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Advanced beaconcha.in graffitiwall image drawer.')
    parser.add_argument('--out-file', default='./graffiti.txt',
                        help='Out location of the generated graffiti file (default: ./graffiti.txt).')
    parser.add_argument('--settings-file', default='./settings.ini',
                        help='Settings file location (default: ./settings.ini).')
    parser.add_argument('--client', required=True, choices=['prysm', 'lighthouse', 'teku', 'nimbus'],
                        help='your eth2 client.')
    parser.add_argument('--eth2-url', default='localhost',
                        help='Your nimbus client url.')  # TODO rename to nimbus/rpc
    parser.add_argument('--eth2-port', default=5052, help='Your nimbus client rest api port.')
    parser.add_argument('--update-wall-time', default=600,
                        help='Interval between graffiti wall updates (default: 600s).')
    parser.add_argument('--update-file-time', default=60, help='Interval between graffiti file updates (default: 60s).')
    args = parser.parse_args()

    config = configparser.ConfigParser()
    config.read(args.settings_file)
    cfg = config['GraffitiConfig']
    img, x_offset, y_offset = getImage()

    white_pixels = np.all(img[:, :, :3] == [255, 255, 255], axis=-1)
    transparent_pixels = img[..., 3] == 0
    static_draw = ~(white_pixels + transparent_pixels)

    last_wall_update = 0
    last_file_update = 0
    wall = dict()
    # file formatting
    pre = ""
    post = ""
    if args.client == "lighthouse":
        pre = "default: "
    elif args.client == "prysm":
        pre = 'ordered:\n  - "'
        post = '"'
    # no pre/post for teku, and nimbus uses rpc anyways
    print("Generating graffitis...")
    while True:
        now = time.time()
        if last_wall_update + args.update_wall_time < now:
            getPixelWallData()
            draw_pixels = updateDrawPixels()
            last_wall_update = now
        if last_file_update + args.update_file_time < now:
            # max 32 bytes/characters.
            # actual graffiti is 15, "RP-X" + spaces + parentheses are 8
            # so we've 9 characters left for version
            graffiti = "RP-" + args.client[0:1].upper() + " " + os.environ['ROCKET_POOL_VERSION'][0:9] + " (" + getPixel() + ")"
            now_string = '[' + str(datetime.now()) + ']: '
            try:
                if args.client == "nimbus":
                    if not setNimbusGraffiti(graffiti):
                        raise Exception("RequestException on calling RPC")
                else:
                    with open(args.out_file, 'w') as f:
                        f.write(pre + graffiti + post)
            except Exception as e:
                print(now_string + 'Error setting graffiti: ', e)
            else:
                print(now_string + 'Graffiti set: ' + graffiti)
            last_file_update = now
        time.sleep(10)

File: drawer.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"github.com/fsnotify/fsnotify"
	"github.com/go-co-op/gocron"
	"github.com/namsral/flag"
	"io"
	"log"
	"math/rand"
	"net/http"
	"os"
	"os/signal"
	"regexp"
	"strings"
	"syscall"
	"time"
)

type GraffitiWallPixel struct {
	Color     string
	Slot      uint
	Validator uint
	X         uint
	Y         uint
}

type GraffitiWall struct {
	Status string
	Data   []GraffitiWallPixel
}

type DrawerPixel struct {
	X     uint   `json: x`
	Y     uint   `json: y`
	Color string `json: color`
}

type DrawerData struct {
	Data []DrawerPixel
}

var (
	OutputFile      string
	InputURL        string
	InputIsURL      bool
	ConsensusClient string
	NimbusURL       string
	Network         string
	UpdateWallTime  int
	UpdateInputTime int
	UpdatePixelTime int
	BeaconURL       string
	GraffitiPrefix  string
	wallCache       *GraffitiWall
	inputCache      *DrawerData
	todoPixels      *DrawerData
	dataChanged     bool
)

func getJson(url string, target interface{}) error {
	log.Printf("Fetching %s for json result....\n", url)
	client := &http.Client{Timeout: 10 * time.Second}
	r, err := client.Get(url)
	if err != nil {
		return err
	}
	defer r.Body.Close()
	return json.NewDecoder(r.Body).Decode(target)
}

func fetchGraffitiWall() bool {
	wallCache = &GraffitiWall{}

	err := getJson(BeaconURL, wallCache)
	if err != nil {
		log.Fatalln(err)
	}
	if wallCache.Status != "OK" {
		log.Printf("WARNING: Unpexpected beaconcha.in graffitiwall status: %s\n", wallCache.Status)
	} else {
		log.Printf("Graffitiwall fetched from beaconcha.in, status: %s\n", wallCache.Status)
	}

	return true
}

func init() {
	flag.StringVar(&OutputFile, "output_file", "/graffiti/graffiti.txt", "File to write graffiti to")
	flag.StringVar(&InputURL, "input_url", "/graffiti/imagedata.json", "URL or path to file to source pixeldata from")
	flag.StringVar(&ConsensusClient, "consensus_client", "", "teku, nimbus, lighthouse, prysm")
	flag.StringVar(&NimbusURL, "nimbus_url", "", "URL to Nimbus (if that's your selected client), e.g. http://127.0.0.1:5052")
	flag.StringVar(&GraffitiPrefix, "graffiti_prefix", "", "Optional graffiti prefix, e.g. 'RPL-L v1.2.3'. Not for Gnosis network (yet).")
	flag.StringVar(&Network, "network", "mainnet", "Network to draw own, e.g. mainnet, gnosis, prater")
	flag.IntVar(&UpdateWallTime, "update_wall_time", 600, "Time between beaconcha.in updates")
	flag.IntVar(&UpdateInputTime, "update_input_time", 600, "Time between input updates - only if remote URL is used. File will be instantly reloaded when changed.")
	flag.IntVar(&UpdatePixelTime, "update_pixel_time", 60, "Time between output updates")
	flag.Parse()

	rand.Seed(time.Now().UnixNano())

	// Sanity
	ccre := regexp.MustCompile(`^(teku|nimbus|lighthouse|prysm)$`)
	ConsensusClient = strings.TrimSpace(strings.ToLower(ConsensusClient))
	if len(ConsensusClient) == 0 || !ccre.MatchString(ConsensusClient) {
		log.Fatalf("Error: invalid/no consensus client supplied, accepted values are: teku nimbus lighthouse prysm")
		return
	}
	if len(GraffitiPrefix) > 16 {
		log.Printf("WARNING: graffiti prefix is too long, will be truncated to [%.16s]!", GraffitiPrefix)
	}
	if Network == "gnosis" && len(GraffitiPrefix) > 0 {
		log.Printf("WARNING: graffiti prefix is disabled for Gnosis right now!")
	}
	if UpdatePixelTime <= 0 {
		UpdatePixelTime = 60
	}
	if UpdateWallTime <= 0 {
		UpdateWallTime = 600
	}
	if UpdateInputTime <= 0 {
		UpdateInputTime = 600
	}

	// Nimbus?
	if ConsensusClient == "nimbus" {
		if len(NimbusURL) == 0 {
			log.Fatalf("Error: nimbus selected, but no nimbus URL supplied.")
			return
		}
	}

	switch Network {
	case "mainnet":
		BeaconURL = "https://beaconcha.in/api/v1/graffitiwall"
	case "gnosis":
		BeaconURL = "https://beacon.gnosischain.com/api/v1/graffitiwall"
	default:
		BeaconURL = fmt.Sprintf("https://%s.beaconcha.in/api/v1/graffitiwall", Network)
	}

	// Input remote or file?
	re := regexp.MustCompile(`^https?://`)
	if re.MatchString(InputURL) {
		InputIsURL = true
		updateInput()
	} else {
		// Make sure input exists and can be loaded
		if fi, err := os.Stat(InputURL); err != nil {
			log.Fatalf("Input file %s does not exist! %s\n", InputURL, err)
		} else {
			log.Printf("Input file %s of size %d found.\n", InputURL, fi.Size())
		}
		if !updateInput() {
			log.Fatalf("Input file %s is not working for us :/\n", InputURL)
		}
	}
}

func updateInput() bool {
	newData := &DrawerData{}
	if InputIsURL {
		err := getJson(InputURL, &newData.Data)
		if err != nil {
			log.Fatalf("ERROR: Could not parse input data from supplied URL [%s]: %s\n", InputURL, err)
		}
	} else {
		inputFile, err := os.Open(InputURL)
		if err != nil {
			log.Fatalf("ERROR: Could not read input for graffiti: %s\nPlease supply either a valid path to input JSON file, OR a valid URL, e.g. https://.\n", err)
		}
		defer inputFile.Close()
		data, err := io.ReadAll(inputFile)
		if err != nil {
			log.Fatalf("ERROR: Could not read input for graffiti: %s\nPlease supply either a valid path to input JSON file, OR a valid URL, e.g. https://.\n", err)
		}
		err = json.Unmarshal(data, &newData.Data)
		if err != nil {
			log.Fatalf("ERROR: Could not parse input for graffiti: %s\nPlease supply either a valid path to input JSON file, OR a valid URL, e.g. https://.\n", err)
		}
	}
	log.Printf("Input now holds %d pixels\n", len(newData.Data))
	inputCache = newData
	dataChanged = true
	return true
}

func updateGraffiti() {
	log.Printf("Updating graffiti...\n")

	// Recalculate todo pixels if anything changed
	if dataChanged || todoPixels == nil {
		// make sure we have a graffiti wall already, otherwise we'll wait till next run
		if wallCache == nil {
			log.Printf("Wall not available yet, waiting...\n")
			return
		}
		if inputCache == nil {
			log.Printf("Input data not available yet, waiting...\n")
			return
		}

		dataChanged = false
		todoPixels = &DrawerData{}

		// Make a lookup dict for walldata
		wallCheck := make(map[string]string)
		for _, pixel := range wallCache.Data {
			// TODO: sanity check on beacon wall data, for now we'll just accept it since we're only comparing against it
			wallCheck[fmt.Sprintf("%s:%s", pixel.X, pixel.Y)] = strings.ToLower(pixel.Color)
		}

		completed := 0
		pcre := regexp.MustCompile(`^[a-f0-9]{6}$`)
		for _, pixel := range inputCache.Data {
			// Is this pixel sane?
			pixel.Color = strings.TrimSpace(strings.ToLower(pixel.Color))
			if pixel.X > 999 || pixel.Y > 999 || !pcre.MatchString(pixel.Color) {
				log.Printf("WARNING: Input json has invalid pixel: %d,%d %s - skipping!\n", pixel.X, pixel.Y, pixel.Color)
				continue
			}
			// Is it already drawn?
			if wallCheck[fmt.Sprintf("%s:%s", pixel.X, pixel.Y)] == strings.ToLower(pixel.Color) {
				completed++
			} else {
				todoPixels.Data = append(todoPixels.Data, pixel)
			}
		}
		log.Printf("STATUS: TODO %d pixel(s), %d already painted!\n", len(todoPixels.Data), completed)
	}

	// grab pixel, generate graffiti, output
	if todoPixels == nil {
		log.Printf("No todoPixels data available yet. Retry soon...\n")
		return
	}
	if len(todoPixels.Data) == 0 {
		log.Printf("Zero todoPixels available... no graffiti for now.... [guard mode]\n")
		return
	}
	index := rand.Intn(len(todoPixels.Data))
	pixel := todoPixels.Data[index]
	log.Printf(" * Randomly selected todo pixel: [%d,%d] #%s\n", pixel.X, pixel.Y, pixel.Color)

	// Format graffiti and send it out
	pre := ""
	post := ""
	switch ConsensusClient {
	case "lighthouse":
		pre = `default: `
	case "prysm":
		pre = `ordered:\n  - "`
		post = `"`
	}

	graffiti := getGraffiti(pixel, Network != "gnosis")

	if Network != "gnosis" && len(GraffitiPrefix) > 0 {
		// pixel is 15, so we've 17 left
		graffiti = strings.TrimSpace(fmt.Sprintf("%.16s %s", GraffitiPrefix, graffiti))
	}
	graffiti = fmt.Sprintf("%s%s%s", pre, graffiti, post)

	if ConsensusClient == "nimbus" {
		err := setNimbusGraffiti(graffiti)
		if err != nil {
			log.Printf(" ❌ WARNING: Could not set Nimbus graffiti to %s: %s\n", graffiti, err)
			return
		}
	} else {
		tmp := []byte(fmt.Sprintf("%s\n", graffiti))
		err := os.WriteFile(OutputFile, tmp, 0644)
		if err != nil {
			log.Printf(" ❌ WARNING: Could not write graffiti %s to %s: %s\n", graffiti, OutputFile, err)
			return
		}
	}
	log.Printf(" ✅ graffiti update OK.\n")
}

func getGraffiti(pixel DrawerPixel, newFormat bool) string {
	if newFormat {
		return fmt.Sprintf("gw:%03d%03d%s", pixel.X, pixel.Y, pixel.Color)
	} else {
		return fmt.Sprintf("graffitiwall:%d:%d:%s", pixel.X, pixel.Y, pixel.Color)
	}
}

func setNimbusGraffiti(graffiti string) error {
	// We send the graffiti data directly to nimbus' api at NIMBUSURL/api/nimbus/v1/graffiti
	url := fmt.Sprintf("%s%s", strings.TrimSuffix(NimbusURL, "/"), "/api/nimbus/v1/graffiti")
	log.Printf("Sending graffiti [%s] to nimbus at %s...\n", graffiti, url)

	// response = requests.post(url, headers=header, data=graffiti)
	client := &http.Client{Timeout: 10 * time.Second}
	r, err := client.Post(url, "text/plain", bytes.NewBufferString(graffiti))
	if err != nil {
		return err
	}
	defer r.Body.Close()
	reply, _ := io.ReadAll(r.Body)
	log.Printf("Server at %s accepted our graffiti, response code: %d, response body of %d bytes\n", url, r.StatusCode, len(reply))
	return nil
}

func writeWatcher(watcher *fsnotify.Watcher) {
	log.Printf("File watcher starting....")
	isChanged, isRemoved := false, false
	for {
		select {
		case event, ok := <-watcher.Events:
			if !ok {
				log.Printf("Watcher not ok??? %s\n", event)
				return
			}
			// log.Printf("Watcher event: %s\n", event)
			if event.Op&fsnotify.Remove == fsnotify.Remove {
				isChanged, isRemoved = true, true
			}
			if event.Op&fsnotify.Write == fsnotify.Write {
				log.Printf("Input file %s modified (%s)!\n", InputURL, event.Name)
				isChanged = true
			}
		case err, ok := <-watcher.Errors:
			if !ok {
				log.Printf("Watcher not ok??? %s\n", err)
				return
			}
			log.Printf("Watcher error: %s\n", err)
		}
		if _, err := os.Stat(InputURL); !os.IsNotExist(err) && (isRemoved || isChanged) {
			if isRemoved {
				log.Printf("Rewatching InputFile %s\n", InputURL)
				err := watcher.Add(InputURL)
				if err != nil {
					log.Printf("ERROR adding InputFile %s to watcher: %s", InputURL, err)
				}
				isChanged, isRemoved = true, false
			}
			if isChanged {
				updateInput()
				isChanged = false
			}
		}
	}
}

func main() {
	sigs := make(chan os.Signal, 1)
	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
	done := make(chan bool, 1)

	log.Printf(" * Drawer starting for %s, generating output every %d seconds.\n", ConsensusClient, UpdatePixelTime)
	cron := gocron.NewScheduler(time.UTC)

	// Cronjob to pull input URL every UpdateInputTime, OR set a file watcher
	if InputIsURL {
		cron.Every(UpdateInputTime).Seconds().Do(updateInput)
	} else {
		// Set watcher
		watcher, err := fsnotify.NewWatcher()
		if err != nil {
			log.Printf("WARNING: Failed to create input watcher for file, reverting to cron updates: ", err)
			cron.Every(UpdateInputTime).Seconds().Do(updateInput)
		} else {
			go writeWatcher(watcher)
			err = watcher.Add(InputURL)
			if err != nil {
				log.Fatalf("ERROR adding InputFile %s to watcher: %s", InputURL, err)
			}
		}
		defer watcher.Close()
		log.Printf("File Watcher on %s installed.\n", InputURL)
	}

	cron.Every(UpdateWallTime).Seconds().Do(fetchGraffitiWall)
	// Slight delay so we hopefully have all data before generation
	cron.Every(UpdatePixelTime).Seconds().StartAt(time.Now().Add(time.Duration(5 * time.Second))).Do(updateGraffiti)

	go func() {
		sig := <-sigs
		log.Printf("%s signal caught, bye bye!\n", sig)
		done <- true
	}()

	cron.StartAsync()
	<-done
}
